language: Update block_comment and documentation comment

claytonrcarter opened the pull request:
As suggested in https://github.com/zed-industries/zed/pull/34418, this proposes various changes to language configs to make block comments and doc-block-style comments more similar. In doing so, it introduces some breaking changes into the extension schema.

This change is needed to support the changes I'm working on in #34418, to be able to support `rewrap` in block comments like `/* really long comment ... */`. As is, we can do this in C-style doc-block comments (eg `/** ... */`) because of the config in `documentation`, but we can't do this in regular block comments because we lack the info about what the line prefix and indentation should be.

And while I was here, I did various other clean-ups, many of which feel nice but are optional.

I would love special attention on the changes to the schema, version and related changes; I'm totally unfamiliar with that part of Zed.

**Summary of changes**
- break: changes type of `block_comment` to same type as `documentation_comment` (**this is the important change**)
- break: rename `documentation` to `documentation_comment` (optional, but improves consistency w/ `line_comments` and `block_comment`)
- break/refactor?: removes some whitespace in the declaration of `block_comment` delimiters (optional, may break things, need input; some langs had no spaces, others did)
- refactor: change `tab_size` from `NonZeroU32` to just a `u32` (some block comments don't seem to need/want indent past the initial delimiter, so we need this be 0 sometimes)
- refactor: moves the `documentation_comment` declarations to appear next to `block_comment`, rearranges the order of the fields in the TOML for `documentation_comment`, rename backing `struct` (all optional)

**Future scope**
I believe that this will also allow us to extend regular block comments on newline – as we do doc-block comments – but I haven't looked into this yet. (eg, in JS try pressing enter in both of these: `/*  */` and `/**  */`; the latter should extend w/ a `*` prefixed line, while the former does not.)

Release Notes:

- N/A

/cc @smitbarmase

maxdeviant reviewed (dismissed):
I think we can make this change in a backwards-compatible way.

We should be able to make it so the language `config.toml` can deserialize both the old and the new format and then migrate old -> new.

claytonrcarter commented:
> We should be able to make it so the language config.toml can deserialize both the old and the new format and then migrate old -> new.

@maxdeviant That makes sense and I considered that during initial implementation, but didn't know where to start. Sorry to be needy, but you have suggestions for where to get started? Maybe recollections of other config items that were updated in a backwards-compatible way? Failing that, would I approach this by writing my own deserializing fn to pass to serde via `deserialize_with`? I'm happy to dig in to this, but I'm totally unfamiliar with this part of the app. Thank you!

smitbarmase commented:
@claytonrcarter I think writing a deserializer is way to go if we want it backwards-compatible, something like this (haven't checked if this works):

```rs
impl<'de> Deserialize<'de> for BlockCommentConfig {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        #[derive(Deserialize)]
        #[serde(untagged)]
        enum BlockCommentConfigHelper {
            New {
                start: Arc<str>,
                end: Arc<str>,
                prefix: Arc<str>,
                tab_size: u32,
            },
            Old([Arc<str>; 2]),
        }

        match BlockCommentConfigHelper::deserialize(deserializer)? {
            BlockCommentConfigHelper::New {
                start,
                end,
                prefix,
                tab_size,
            } => Ok(BlockCommentConfig {
                start,
                end,
                prefix,
                tab_size,
            }),
            BlockCommentConfigHelper::Old([start, end]) => Ok(BlockCommentConfig {
                start,
                end,
                prefix: "".into(),
                tab_size: 0,
            }),
        }
    }
}
```

claytonrcarter commented:
Thank you for the hint, @smitbarmase and @maxdeviant. I can take that example and run with it.

smitbarmase commented on crates/language/src/language.rs:
```diff
@@ -2460,4 +2514,59 @@ mod tests {
         );
     }
+
+    #[test]
```
Thanks for testing this. I believe it’s safe to remove this now that we’ve confirmed it works. I don’t want to add an additional dependency, and this configuration is unlikely to change frequently.

maxdeviant replied:
I think we should keep the test, but we can move the dependency to a dev dependency: https://github.com/zed-industries/zed/pull/34861#discussion_r2225449983

claytonrcarter replied:
I kept it and updated it to test the new and old formats/names, and moved the dep to a dev dep.

smitbarmase commented on crates/language/src/language.rs:
```diff
@@ -731,7 +731,7 @@ pub struct LanguageConfig {
     pub block_comment: Option<BlockCommentConfig>,
     /// Delimiters and configuration for recognizing and formatting documentation comments.
-    #[serde(default)]
+    #[serde(default, deserialize_with = "deserialize_doc_comment", flatten)]
```
Thanks for considering this! Though I don't think an extension maintainer would add both in the config without testing if it works, so it probably can be simplified to use an alias.  Lesser the code, better 😄.

smitbarmase reviewed (changes requested):
It looks great to me. I’ve added a few clean-up / nit-pick comments. Once those are addressed, it should be ready to merge.

claytonrcarter committed:
compat: make config changes backward compatible

claytonrcarter committed:
feat(language)!: make block_comment similar to documentation_comment

claytonrcarter committed:
refactor(languages): update block/doc comments in languages

- update syntax of block_comment
- rename documentation_comment
- move documentation_comment next to block_comment
- misc minor updates

claytonrcarter commented on crates/extension_host/src/extension_host.rs:
```diff
@@ -70,7 +70,7 @@ pub const RELOAD_DEBOUNCE_DURATION: Duration = Duration::from_millis(200);
 
 /// The current extension [`SchemaVersion`] supported by Zed.
-const CURRENT_SCHEMA_VERSION: SchemaVersion = SchemaVersion(1);
+const CURRENT_SCHEMA_VERSION: SchemaVersion = SchemaVersion(2);
```
OK, reverted, as well as the schema bumps in the 2 extensions.

smitbarmase commented:
Thank you for working on this!
